src/main/java/org/truffleruby/language/yield/YieldExpressionNode.java

```diff
@@ -51,13 +53,24 @@ public final Object execute(VirtualFrame frame) {
             useCapturedBlock.enter();
 
             block = RubyArguments.getMethod(frame).getCapturedBlock();
+        }
+
+        if (block == null) {
+            // TODO (pitr-ch 19-Dec-2017): a hack to get zlib working, needs proper investigation
+            // captured block passing down is messy, it abuses blockArgument which is suppose to be left null
```

Comment:
What is `blockArgument`?

Reply:
RubyArguments.ArgumentIndicies#BLOCK

Reply:
I don't understand the second comment.
Is it about the `if` at line 50?
The block argument is also set in blocks frames currently, not just in method frames.

Reply:
This is a **hack** and I'll be looking for a proper fix soon. This hack works for usual zlib execution and specs, but it's not correct and will break in other corner cases.
The comment about block argument is related to following problem:
```ruby
def a_method
  [1].each { |*a, &block| puts block.call }
end
a_method { 2 } # prints 2
```
which looks like it'll need to be fixed for yield to work properly, therefore it's noted in the same todo.

Reply:
I see. Could you add that piece of code to the comment to clarify?
Or even better write a language spec and mention the spec file here?

Reply:
Yes I had already planned to do it in a followup PR.